### ISSUE TITLE
SVG with large amount of linked images crashing

### ISSUE THREAD

**Issue by CapNBritain** [exupero/saveSvgAsPng#77]
Hello, I am new to GitHub so I apologize in advance if I don't format my issue properly or follow the correct conventions.  I am making a site that lets users build SVG maps and want to add Save as Image functionality.  I have tried to use your code, but the browser if the SVG becomes too complex.  I have two test pages set up.

This one is small and works:
http://fantasymapbuilder.com/test-save-small.html

This one is larger and crashes:
http://fantasymapbuilder.com/test-save.html

I have made a few small changes to your code to get it to work on smaller maps, but I think the only one that matters is in function inlineImages where I check if href is blank or null.  This is because I set SVG image href attributes to an empty string when I want to blank them out.

``` javascript
if (href != '' && href != null) {
    img.setAttribute('crossOrigin', 'anonymous');
    img.onload = function() {
      canvas.width = img.width;
      canvas.height = img.height;
      ctx.drawImage(img, 0, 0);
      image.setAttributeNS("http://www.w3.org/1999/xlink", "href", canvas.toDataURL('image/png'));
      left--;
      if (left == 0) {
        callback();
      }
    }
    img.onerror = function() {
      console.log("Could not load "+href);
      left--;
      if (left == 0) {
        callback();
      }
    }
    img.src = href;
} else {
    left--;
    if (left == 0) {
      callback();
    }
}
```

Do you have any thoughts on why the browser is crashing on the large test case?  Is it simply a matter of it not being able to handle so many images?  Am I completely out of luck in trying to save my complex SVGs as images?  Thanks in advance.

**Comment by exupero** [comment]
I've added your additional check for an href to the latest version.

Both of those links do, in fact, download the image for me, though they are very slow and on the larger one I have to tell Firefox not to stop the script. I suspect that with so many images linked in the SVG, the inlining process takes a very long time.

To speed it up, you could use base64-encoded images embedded in the SVG, rather than linked images. This will bypass the inlining process (which just turns linked images into based64-encoded embedded ones) and will make the download faster.

**Comment by CapNBritain** [comment]
Thank you for your help.  I followed your advice and decided to base64-encode all my images.  I ripped out the meat of your inlineImages function and used an AJAX call to a php script that does it for me instead of loading the images into an Image object.  Due to the size of my svg it can still crash during the toDataUrl step, but I'm finding ways to mitigate that.  Thanks for pointing me in the right direction.
